comctl32.go

```diff
@@ -388,3 +394,191 @@ func LoadIconWithScaleDown(hInstance HINSTANCE, lpIconName *uint16, w int32, h i
+)
+
+const (
+	TDN_CREATED                = 0
+	TDN_NAVIGATED              = 1
+	TDN_BUTTON_CLICKED         = 2
+	TDN_HYPERLINK_CLICKED      = 3
+	TDN_TIMER                  = 4
+	TDN_DESTROYED              = 5
+	TDN_RADIO_BUTTON_CLICKED   = 6
+	TDN_DIALOG_CONSTRUCTED     = 7
+	TDN_VERIFICATION_CLICKED   = 8
+	TDN_HELP                   = 9
+	TDN_EXPANDO_BUTTON_CLICKED = 10
+)
+
+// TASKDIALOG_BUTTON_UNPACKED is nearly identical to TASKDIALOG_BUTTON in the Windows
+// SDK, except for the fact that that the SDK version is packed. Since Go cannot
+// grok that, we implement a Pack method on this struct to encode its contents
+// correctly for Windows.
+type TASKDIALOG_BUTTON_UNPACKED struct {
+	ButtonID   int32
+	ButtonText *uint16
+}
+
+// TASKDIALOG_BUTTON is opaque because it is a packed data structure.
+// Use (*TASKDIALOG_BUTTON_UNPACKED).Pack() to write one to a bytes.Buffer.
+type TASKDIALOG_BUTTON struct {
+	_ [sizeTASKDIALOG_BUTTON]byte
+}
+
+// Pack writes the contents of u to buf in a packed format. We use a Buffer
+// because walk encodes TASKDIALOG_BUTTONs from a slice.
+func (u *TASKDIALOG_BUTTON_UNPACKED) Pack(buf *bytes.Buffer) error {
+	if err := binary.Write(buf, packByteOrder, u.ButtonID); err != nil {
+		return err
+	}
+	return binary.Write(buf, packByteOrder, packPtr(uintptr(unsafe.Pointer(u.ButtonText))))
+}
+
+type TASKDIALOG_ELEMENTS int32
+
+const (
+	TDE_CONTENT              = TASKDIALOG_ELEMENTS(0)
+	TDE_EXPANDED_INFORMATION = TASKDIALOG_ELEMENTS(1)
+	TDE_FOOTER               = TASKDIALOG_ELEMENTS(2)
+	TDE_MAIN_INSTRUCTION     = TASKDIALOG_ELEMENTS(3)
+)
+
+type TASKDIALOG_ICON_ELEMENTS int32
+
+const (
+	TDIE_ICON_MAIN   = TASKDIALOG_ICON_ELEMENTS(0)
+	TDIE_ICON_FOOTER = TASKDIALOG_ICON_ELEMENTS(1)
+)
+
+const (
+	TD_WARNING_ICON     = uintptr(0xFFFF)
+	TD_ERROR_ICON       = uintptr(0xFFFE)
+	TD_INFORMATION_ICON = uintptr(0xFFFD)
+	TD_SHIELD_ICON      = uintptr(0xFFFC)
+)
+
+type TASKDIALOG_COMMON_BUTTON_FLAGS uint32
+
+const (
+	TDCBF_OK_BUTTON     = TASKDIALOG_COMMON_BUTTON_FLAGS(0x0001)
+	TDCBF_YES_BUTTON    = TASKDIALOG_COMMON_BUTTON_FLAGS(0x0002)
+	TDCBF_NO_BUTTON     = TASKDIALOG_COMMON_BUTTON_FLAGS(0x0004)
+	TDCBF_CANCEL_BUTTON = TASKDIALOG_COMMON_BUTTON_FLAGS(0x0008)
+	TDCBF_RETRY_BUTTON  = TASKDIALOG_COMMON_BUTTON_FLAGS(0x0010)
+	TDCBF_CLOSE_BUTTON  = TASKDIALOG_COMMON_BUTTON_FLAGS(0x0020)
+)
+
+// TASKDIALOGCONFIG_UNPACKED is nearly identical to TASKDIALOGCONFIG in the Windows
+// SDK, except for the fact that that the SDK version is packed. Since Go cannot
+// grok that, we implement a Pack method on this struct to encode its contents
+// correctly for Windows.
+type TASKDIALOGCONFIG_UNPACKED struct {
+	Size                 uint32 // Size is included for consistency with the SDK but is ignored by Pack()
+	HWNDParent           HWND
+	HInstance            HINSTANCE
+	Flags                TASKDIALOG_FLAGS
+	CommonButtons        TASKDIALOG_COMMON_BUTTON_FLAGS
+	WindowTitle          *uint16
```

Comment:
Is `PCWSTR` as `*uint16` a standard convention here? Maybe there should be some other type to make it easier to manipulate? Obviously, not in this PR.

Reply:
Not used previously in this codebase... I'm also a bit wary of doing that in some cases, eg some of these hungarian type names imply things that can't be done in go (eg `PCWSTR` implies `const`). Something to consider how to do in the future, for sure.